hivemq-edge/src/frontend/src/extensions/datahub/designer/data_policy/DataPolicyNode.utils.ts

```diff
@@ -0,0 +1,114 @@
+import { getIncomers, getOutgoers, Node } from 'reactflow'
+
+import { DataPolicy, DataPolicyValidator, PolicyOperation } from '@/api/__generated__'
+import { DataHubNodeType, DataPolicyData, DryRunResults, TopicFilterData, WorkspaceState } from '@datahub/types.ts'
+import { PolicyCheckErrors } from '@datahub/designer/validation.errors.ts'
+
+export function getSubFlow(source: Node, acc: Node[], store: WorkspaceState, only = false) {
+  const allIds = Array.from(new Set(acc.map((e) => e.id)))
+  const { nodes, edges } = store
+
+  // This is wrong: should check it's from the right handles
+  const incomers = getIncomers(source, nodes, edges).filter(
+    (node) =>
+      !allIds.includes(node.id) &&
+      node.type !== DataHubNodeType.DATA_POLICY &&
+      node.type !== DataHubNodeType.BEHAVIOR_POLICY
+  )
+  const outgoers = getOutgoers(source, nodes, edges).filter(
+    (node) =>
+      !allIds.includes(node.id) &&
+      node.type !== DataHubNodeType.DATA_POLICY &&
+      node.type !== DataHubNodeType.BEHAVIOR_POLICY
+  )
+
+  acc.push(...incomers, ...outgoers)
+
+  if (outgoers.length) {
+    outgoers.forEach((node) => {
+      const subFlow = getSubFlow(node, acc, store)
+      acc = Array.from(new Set([...acc, ...subFlow]))
+    })
+  }
+
+  if (incomers.length && !only) {
+    incomers.forEach((node) => {
+      const subFlow = getSubFlow(node, acc, store)
+      acc = Array.from(new Set([...acc, ...subFlow]))
+    })
+  }
+
+  return acc
+}
```

Comment:
question: this one is really heavy, if you like we can go trough that one in our paring and simply it.

Reply:
https://hivemq.kanbanize.com/ctrl_board/57/cards/19805/details/